exploration/0474-text-vs-code.md

```diff
@@ -0,0 +1,366 @@
+This is often related to plural cases, such as "You have 3 new messages",
+where the value `3` is an input and the "messages" needs to correspond with its plural category.
+
+Rarely, messages needs to include leading or trailing whitespace due to
+e.g. how they will be concatenated with other text,
+or as a result of being segmented from some larger volume of text.
+
+## Requirements
+
+Easy things should be easy, and hard things should be possible.
+
+Developers and translators should be able to read and write the syntax easily in a text editor.
+
+As MessageFormat 2 will be at best a secondary language to all its users,
+it should conform to user expectations and require as little learning as possible.
+
+The syntax should avoid footguns,
+in particular as it's passed through various tools during formatting.
+
+## Constraints
+
+Limiting the range of characters that need to be escaped in plain text is important.
+Following past precedent,
+this design doc will only consider encapsulation styles which
+start with `{` and end with `}`.
+
+The current syntax includes some plain-ascii keywords:
+`input`, `local`, `match`, and `when`.
+
+The current syntax and active proposals include some sigil + name combinations,
+such as `:number`, `$var`, `|literal|`, `+bold`, and `@attr`.
+
+The current syntax supports unquoted literal values as operands.
+
+Messages themselves are "simple strings" and must be considered to be a single
+line of text. In many containing formats, newlines will be represented as the local
+equivalent of `\n`.
+
+## Proposed Design
+
+TBD
+
+## Alternatives Considered
+
+### Start in code, encapsulate text
+
+This approach treats messages as something like a resource format for pattern values.
+Keywords are declared directly at the top level of a message,
+and patterns are surrounded by `{...}`.
+
+Whitespace in patterns is never trimmed.
+
+Some code statements (variable declarations and match statements)
+also use `{...}` to surround values at the top level,
+so counting `{` instances is not sufficient to identify if a value is "code" or "text".
+
+The `{...}` are required for all messages,
+including ones that only consist of text.
+Delimiters of the resource format are required in addition to this,
+so messages may appear wrapped as e.g. `"{...}"`.
+
+Examples:
+
+```
+{Hello world}
+```
+
+```
+{Hello {$user}}
+```
+
+```
+input {$count :number minimumFractionDigits=1}
+{You have eaten {$count} apples}
+```
+
+```
+input {$count :number}
+match {$count}
+when 0 {You have no new message}
+when one {You have {$count} new message}
+when * {You have {$count} new messages}
+```
+
+```
+{ and some more}
+```
+
+### Start in text, encapsulate code
+
+The approach treats messages as template strings,
+which may include statements and expressions surrounded by `{...}`.
+Multi-variant messages require `match` and `when` statements that are followed by text at the top level.
+
+Whitespace around statements may need to be trimmed
```

Comment:
Trimming spaces are not the only problematic area.
New lines are, but leading / trailing, and in the middle of the text.

Because this syntax is intended to be "embedded" in existing file formats (xml, json, .properties, templating, othwers)